nextcloud/server
# fix(session): Make session encryption more robust

* Resolves: https://github.com/nextcloud/server/issues/42157

## Summary

Move encryption/decryption from a wrapper class to a PHP session handler.

## TODO

- [x] Make it work for new installations
- [x] Make it work for upgrades
- [x] Encrypt/decrypt
- [ ] Testing, testing, testing
- [x] Find a good approach for ``\OCP\ISession::getId``
- [x] BUG: Infinite redirect loop at logout due to `clearingExecutionContexts` session key not clearing
- [x] BUG: https://github.com/nextcloud/twofactor_webauthn/issues/713

## How to test - new session

1. Clean your sessions (restart web server for local session file handler, clear Redis for redis, etc)
2. Clear your cookies for the dev setup domain
3. Open Nextcloud
4. Log in with the browser network tools open

There will be a cookie named like the instanceid of your setup. The value is a combination of a shorter string, `|` and a longer string. The short string is the internal session ID. If you look into the local session file or Redis the session data will be encrypted binary data.

## How to test - upgrade

1. Check out master
2. Clean your sessions (restart web server for local session file handler, clear Redis for redis, etc)
3. Clear your cookies for the dev setup domain
4. Open Nextcloud
5. Log in

There will be a cookie named like the instanceid of your setup and a oc_sessionPassphrase cookie. The first cookie will have a relatively short string. f you look into the local session file or Redis the session data will not be encrypted. You should see one key `encrypted_session_data` with an *encrypted value*.

7. Switch to this branch
8. Reload the page with the browser network tools open

The cookies will stay the same. But the session file changes. You will no longer see the `encrypted_session_data` because all session data is encrypted.
Nextcloud continues to work like before.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
